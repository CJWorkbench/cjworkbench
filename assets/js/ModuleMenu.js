// A menu to select a module addition
import React from 'react'
import { UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem } from 'reactstrap'
import PropTypes from 'prop-types'

export default class ModuleMenu extends React.Component {
  constructor(props) {
    super(props);
    this.itemClick = this.itemClick.bind(this);
    this.state = {
      items: []
    };
  }

  componentDidMount() {
    var _this = this;
    fetch('/api/modules/', { credentials: 'include'})
      .then(response => response.json())
      .then(json => {
        // Sort modules first by category, then name
        json.sort( (a,b) => {
          if (a.category > b.category)
            return 1;
          else if (a.category < b.category)
            return -1;
          else if (a.name > b.name)
            return 1;
          else if (a.name < b.name)
            return -1;
          else
            return 0;
        } );
        _this.setState({dropdownOpen: this.state.dropdownOpen, items: json}) })
  }

  itemClick(evt) {
    var itemID = evt.target.getAttribute('data-id');
    this.props.addModule(itemID);
  }

  render() {

    // Construct list of modules, separated by category
    var menuItems = [];
    var lastCategory = undefined;
    for (var item of this.state.items) {
      if (item.category != lastCategory) {
        if (lastCategory)
          menuItems.push(<DropdownItem divider key={item.id*1000} />);
        menuItems.push(<DropdownItem header key={item.id*1000 + 1}> {item.category} </DropdownItem>);
        lastCategory = item.category;
      }
      menuItems.push(<DropdownItem className='ml-2' key={item.id} data-id={item.id} onClick={this.itemClick}> {item.name} </DropdownItem>);
    }

    return (
      <UncontrolledDropdown>
        <DropdownToggle caret>
          Add Module
        </DropdownToggle>
        <DropdownMenu>
          {menuItems}
        </DropdownMenu>
      </UncontrolledDropdown>
    );
  }
}

ModuleMenu.propTypes = {
  addModule:  PropTypes.func,
};
