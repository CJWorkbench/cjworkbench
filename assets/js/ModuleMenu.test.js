import React from 'react';
import { shallow, render, mount } from 'enzyme';

import ModuleMenu from './ModuleMenu';

it('renders correctly', () => {

  var mockModules = [
    {
      "id": 9,
      "name": "Cast",
      "category": "Wrangle"
    },
    {
      "id": 6,
      "name": "Chart",
      "category": "Visualize"
    },
    {
      "id": 1,
      "name": "Formula",
      "category": "Analyze"
    },
    {
      "id": 2,
      "name": "Load From URL",
      "category": "Sources"
    },
    {
      "id": 10,
      "name": "Melt",
      "category": "Wrangle"
    }
    ];


  const wrapper = shallow( <ModuleMenu addModule={ () => {} } /> );
  expect(wrapper).toMatchSnapshot();

  // Now give it some actual menu items
  wrapper.setState( {dropdownOpen: true, items: mockModules} )
  expect(wrapper).toMatchSnapshot();
});

