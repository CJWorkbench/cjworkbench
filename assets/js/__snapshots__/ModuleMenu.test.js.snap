// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Ms
  isOpen={false}
  toggle={[Function]}
>
  <ei
    aria-haspopup={true}
    caret={true}
    color="secondary"
    data-toggle="dropdown"
  >
    Add Module
  </ei>
  <Xo
    tag="div"
  />
</Ms>
`;

exports[`renders correctly 2`] = `
<Ms
  isOpen={true}
  toggle={[Function]}
>
  <ei
    aria-haspopup={true}
    caret={true}
    color="secondary"
    data-toggle="dropdown"
  >
    Add Module
  </ei>
  <Xo
    tag="div"
  >
    <ei
      header={true}
      tag="button"
      toggle={true}
    >
       
      Wrangle
       
    </ei>
    <ei
      className="ml-2"
      data-id={9}
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
       
      Cast
       
    </ei>
    <ei
      divider={true}
      tag="button"
      toggle={true}
    />
    <ei
      header={true}
      tag="button"
      toggle={true}
    >
       
      Visualize
       
    </ei>
    <ei
      className="ml-2"
      data-id={6}
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
       
      Chart
       
    </ei>
    <ei
      divider={true}
      tag="button"
      toggle={true}
    />
    <ei
      header={true}
      tag="button"
      toggle={true}
    >
       
      Analyze
       
    </ei>
    <ei
      className="ml-2"
      data-id={1}
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
       
      Formula
       
    </ei>
    <ei
      divider={true}
      tag="button"
      toggle={true}
    />
    <ei
      header={true}
      tag="button"
      toggle={true}
    >
       
      Sources
       
    </ei>
    <ei
      className="ml-2"
      data-id={2}
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
       
      Load From URL
       
    </ei>
    <ei
      divider={true}
      tag="button"
      toggle={true}
    />
    <ei
      header={true}
      tag="button"
      toggle={true}
    >
       
      Wrangle
       
    </ei>
    <ei
      className="ml-2"
      data-id={10}
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
       
      Melt
       
    </ei>
  </Xo>
</Ms>
`;
